components/eam/src/physics/cam/gw_drag.F90

```diff
@@ -886,14 +895,28 @@ subroutine gw_tend(state, sgh, pbuf, dt, ptend, cam_in)
         ptend%u(:ncol,k) = ptend%u(:ncol,k) + utgw(:,k)
         vtgw(:,k) = vtgw(:,k) * cam_in%landfrac(:ncol)
         ptend%v(:ncol,k) = ptend%v(:ncol,k) + vtgw(:,k)
-        ptend%s(:ncol,k) = ptend%s(:ncol,k) + ttgw(:,k) &
-             -(ptend%u(:ncol,k) * (u(:,k) + ptend%u(:ncol,k)*0.5_r8*dt) &
-             +ptend%v(:ncol,k) * (v(:,k) + ptend%v(:ncol,k)*0.5_r8*dt))
-        ttgw(:,k) = ttgw(:,k) &
-             -(ptend%u(:ncol,k) * (u(:,k) + ptend%u(:ncol,k)*0.5_r8*dt) &
-             +ptend%v(:ncol,k) * (v(:,k) + ptend%v(:ncol,k)*0.5_r8*dt))
-        ttgw(:,k) = ttgw(:,k) / cpairv(:ncol, k, lchnk)
-     end do
+     enddo
+
+     if(.not. use_gw_energy_fix1) then
+        !original code
+        do k = 1, pver
+           ptend%s(:ncol,k) = ptend%s(:ncol,k) + ttgw(:,k) &
+                -(ptend%u(:ncol,k) * (u(:,k) + ptend%u(:ncol,k)*0.5_r8*dt) &
+                +ptend%v(:ncol,k) * (v(:,k) + ptend%v(:ncol,k)*0.5_r8*dt))
+           ttgw(:,k) = ttgw(:,k) &
+                -(ptend%u(:ncol,k) * (u(:,k) + ptend%u(:ncol,k)*0.5_r8*dt) &
+                +ptend%v(:ncol,k) * (v(:,k) + ptend%v(:ncol,k)*0.5_r8*dt))
+           ttgw(:,k) = ttgw(:,k) / cpairv(:ncol, k, lchnk)
+        end do
+     else
+        !new code, brute force to conserve energy in each level locally
```

Comment:
Hi @oksanaguba , compared to original, with fix1, it seems to amount to throwing away temperature tendency ttgw (which is of diffusion of dry static energy and conversion of tke)  from all gw sources. Is this interpretation correct?

Reply:
i am not sure what you mean. the change in this pr overwrites ttgw at the end of gw_tend call but at that point ttgw is not actually used anywhere below (as it is not used in the original). even for diagnostics, ptend%s is used instead:
  call outfld ('TTGW', ptend%s/cpairv(:,:,lchnk),  pcols, lchnk)
also, from what i saw ttgw is always ptend%s/c_p .

assuming ptend%u and ptend%v are given, to conserve energy one needs to modify ptend%s (which modifies temperature later, in the interface). there are a few ways to conserve energy, but if ttgw=ptend%s carries some other meaning/quantity that is needed, please, let me know. thanks.

Reply:
Sorry for not making it clearer. I was focusing on the resulting ptend%s that will be returned to the caller. 

When I mentioned ttgw, I didn't mean the final one. I meant the ttgw returned from each call to gw_drag_pro before this conditional nlock.  In the original codes, ptend%s accumulates these ttgw,   plus -(udu+vdv). With fix1, the accumulation in ptend%s up to this block is not retained.  

ttgw calculated inside gw_drag_pro is the sum of diffusion of dry static energy and conversion of kinetic energy, as the comments and code flow in gw_drag_pro suggest. I have to admit I am not an expert on GWD; just following the codes

Reply:
yes, the original values in ttgw and ptend%s are lost.